processing/src/main/java/io/druid/query/select/SelectQueryQueryToolChest.java

```diff
@@ -243,29 +266,44 @@ public Object apply(final Result<SelectResultValue> input)
 
             DateTime timestamp = granularity.toDateTime(((Number) resultIter.next()).longValue());
 
-            return new Result<SelectResultValue>(
+            List<DimensionSpec> dimensionSpecs = query.getDimensions();
+
+            Map<String, Integer> pageIdentifier = jsonMapper.convertValue(
+                resultIter.next(), new TypeReference<Map<String, Integer>>() {}
+                );
+            Set<String> dimensionSet = jsonMapper.convertValue(
+                resultIter.next(), new TypeReference<Set<String>>() {}
+            );
+            Set<String> metricSet = jsonMapper.convertValue(
+                resultIter.next(), new TypeReference<Set<String>>() {}
+            );
+            List<EventHolder> eventHolders = jsonMapper.convertValue(
+                resultIter.next(), new TypeReference<List<EventHolder>>() {}
+                );
+            if (dimensionSpecs != null) {
+              List<String> cachedOutputNames = (List<String>) resultIter.next();
```

Comment:
This might throw NoSuchElementException if we're pulling a cache value written by an old version of Druid to a shared cache (like memcached). One way we could handle this by adjusting the search query cache key in such a way that old cached values will never be pulled (add some incompatibility to the cache key).